File: src/Shared/Polyrific.Catapult.Shared.ApiClient/JobQueueService.cs
// Copyright (c) Polyrific, Inc 2018. All rights reserved.

using System.Collections.Generic;
using System.Threading.Tasks;
using Polyrific.Catapult.Shared.Dto.JobQueue;
using Polyrific.Catapult.Shared.Service;

namespace Polyrific.Catapult.Shared.ApiClient
{
    public class JobQueueService : BaseService, IJobQueueService
    {
        public JobQueueService(IApiClient api) : base(api)
        {
        }

        public async Task CancelJobQueue(int projectId, int queueId)
        {
            var path = $"project/{projectId}/queue/{queueId}/cancel";

            await Api.Post<object, object>(path, null);
        }

        public async Task<JobDto> CheckJob()
        {
            var path = $"queue";

            return await Api.Get<JobDto>(path);
        }

        public async Task<JobDto> CreateJobQueue(int projectId, NewJobDto newJobQueue)
        {
            var path = $"project/{projectId}/queue";

            return await Api.Post<NewJobDto, JobDto>(path, newJobQueue);
        }

        public async Task<string> GetJobLogs(int projectId, int queueId)
        {
            var path = $"project/{projectId}/queue/{queueId}/logs";

            return await Api.Get<string>(path);
        }

        public async Task<JobDto> GetJobQueue(int projectId, int queueId)
        {
            var path = $"project/{projectId}/queue/{queueId}";

            return await Api.Get<JobDto>(path);
        }

        public async Task<JobDto> GetJobQueue(int projectId, string queueCode)
        {
            var path = $"project/{projectId}/queue/code/{queueCode}";

            return await Api.Get<JobDto>(path);
        }

        public async Task<List<JobDto>> GetJobQueues(int projectId, string filter)
        {
            var path = $"project/{projectId}/queue?filter={filter}";

            return await Api.Get<List<JobDto>>(path);
        }

        public async Task<List<JobTaskStatusDto>> GetJobQueueStatus(int projectId, int queueId, string filter)
        {
            var path = $"project/{projectId}/queue/{queueId}/status?filter={filter}";
            return await Api.Get<List<JobTaskStatusDto>>(path);
        }

        public async Task UpdateJobQueue(int queueId, UpdateJobDto job)
        {
            var path = $"queue/{queueId}";

            await Api.Put(path, job);
        }
    }
}

File: src/Web/opencatapultweb/src/app/core/services/job-queue.service.ts
import { Injectable } from '@angular/core';
import { ApiService } from './api.service';
import { JobDto } from '../models/job-queue/job-dto';
import { NewJobDto } from '../models/job-queue/new-job-dto';
import { SignalrService } from './signalr.service';
import { Observable, Subject } from 'rxjs';
import { JobLogDto } from '../models/job-queue/job-log-dto';

@Injectable()
export class JobQueueService {
  message = new Subject<JobLogDto>();

  constructor(
    private api: ApiService,
    private messageService: SignalrService) { }

  getJobQueues(projectId: number, filter: string) {
    return this.api.get<JobDto[]>(`project/${projectId}/queue?filter=${filter}`);
  }

  getJobQueue(projectId: number, queueId: number) {
    return this.api.get<JobDto>(`project/${projectId}/queue/${queueId}`);
  }

  addJobQueue(projectId: number, dto: NewJobDto) {
    return this.api.post<JobDto>(`project/${projectId}/queue`, dto);
  }

  updateJobQueue(projectId: number, queueId: number, dto: JobDto) {
    return this.api.put(`project/${projectId}/queue/${queueId}`, dto);
  }

  getTaskLogs(projectId: number, queueId: number, taskName: string) {
    return this.api.getString(`project/${projectId}/queue/${queueId}/task/name/${taskName}/logs`);
  }

  listenJobQueueLog(projectId: number, queueId: number): Observable<JobLogDto> {
    this.messageService.connect(`jobQueueHub?projectId=${projectId}&jobQueueId=${queueId}`, this.message);
    return this.message;
  }
}
